framework/web/Request.php

```diff
@@ -1999,6 +2128,86 @@ private function validateCsrfTokenInternal($clientSuppliedToken, $trueToken)
         return $security->unmaskToken($clientSuppliedToken) === $security->unmaskToken($trueToken);
     }
 
+    /**
+     * {@inheritdoc}
+     * @since 2.1.0
+     */
+    public function getAttributes()
+    {
+        if ($this->_attributes === null) {
+            $this->_attributes = $this->defaultAttributes();
+        }
+        return $this->_attributes;
+    }
+
+    /**
+     * @param array $attributes attributes derived from the request.
+     */
+    public function setAttributes(array $attributes)
+    {
+        $this->_attributes = $attributes;
+    }
+
+    /**
+     * {@inheritdoc}
+     * @since 2.1.0
+     */
+    public function getAttribute($name, $default = null)
+    {
+        $attributes = $this->getAttributes();
+        if (!array_key_exists($name, $attributes)) {
+            return $default;
+        }
+
+        return $attributes[$name];
+    }
+
+    /**
+     * {@inheritdoc}
+     * @since 2.1.0
+     */
+    public function withAttribute($name, $value)
+    {
+        $attributes = $this->getAttributes();
+        if (array_key_exists($name, $attributes) && $attributes[$name] === $value) {
+            return $this;
+        }
+
+        $attributes[$name] = $value;
+
+        $newInstance = clone $this;
+        $newInstance->setAttributes($attributes);
+        return $newInstance;
+    }
+
+    /**
+     * {@inheritdoc}
+     * @since 2.1.0
+     */
+    public function withoutAttribute($name)
+    {
+        $attributes = $this->getAttributes();
+        if (!array_key_exists($name, $attributes)) {
+            return $this;
+        }
+
+        unset($attributes[$name]);
+
+        $newInstance = clone $this;
+        $newInstance->setAttributes($attributes);
+        return $newInstance;
+    }
+
+    /**
+     * Returns default server request attributes to be used in case they are not explicitly set.
+     * @return array attributes derived from the request.
+     * @since 2.1.0
+     */
+    protected function defaultAttributes()
```

Comment:
What's the usecase?

Reply:
According to PSR request attributes are determined by particular `Request` class logic.
Having separated method allows developer to define internal attributes while extending `yii\web\Request` without necessity of re-defining private field for thier actual storage.